AAVE activities

James-Sangalli opened the pull request:
Replaces https://github.com/AlphaWallet/TokenScript-Examples/pull/105

This follows the following story:
- User opens ETH card and supplies ETH to AAVE
- User sees new aETH token card and uses it to withdraw ETH or borrow DAI
- After borrowing DAI, user sees DAI card and can repay the DAI debt borrowed against the ETH collateral (requires enabling)

NB: Corresponding activities are created for each part of the story 

Closes #106, #100 #101 

TODO: 
- [x] Show collateral as it relates to the amount of DAI you can borrow
- [x] Show liquidity as it relates to how much ETH you can withdraw after a DAI loan
- [x] Show DAI debt in repay action inside DAI 
- [ ] Test to see that it works properly 
- [ ] implement UI based of design (to be done in another PR after this one is merged) 

FYI @zhangzhongnan928

colourful-land commented:
Why do you underscore every element?

>    `<element name="_fromMinter" ethereum:type="address" ethereum:indexed="true"/>`

You said you prefer these not to be mixed with token attributes anyway:

> I prefer this syntax for the web3 object as it separates them and prevents collisions (while also letting you know what is global and what is based on cards)
````
web3.tokens.dataChanged = (oldTokens, updated, tokenIdCard) => {
    document.getElementById(tokenIdCard).innerHTML = new Token(updated.token, updated.card).render();
};

constructor(token, card) {
        this.token = token;
        this.card = card;
}
````

colourful-land commented:
@hboon check if you want to open an todo in iOS to allow @James-Sangalli's way of addressing native ETH token:
https://github.com/AlphaWallet/TokenScript-Examples/blob/9144a533eb4a55fe2fb3c9734c6576c1130e7000/examples/action/ETH/ETH.xml#L51

James-Sangalli commented:
> @hboon
> 
> ## Special ETH address
> check if you want to open an todo in iOS to allow @James-Sangalli's way of addressing native ETH token:
> https://github.com/AlphaWallet/TokenScript-Examples/blob/9144a533eb4a55fe2fb3c9734c6576c1130e7000/examples/action/ETH/ETH.xml#L51
> 
> @James-Sangalli also used `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` as a special address for Eth native currency, since he used it as a payload to AAVELendingPoolCore's `getReserveCurrentLiquidityRate()`, I assume that's how that contract takes it.
> We will consider introducing his former (000) syntax or latter (Eeee) syntax for native currency (or invent something else) in weekly meetings in Aug, but now this is only needed for the train-leaves branch.

@colourful-land can confirm that the contract accepts the Eeeee address for ETH

James-Sangalli commented:
> Why do you underscore every element?
> 
> > `<element name="_fromMinter" ethereum:type="address" ethereum:indexed="true"/>`
> 
> You said you prefer these not to be mixed with token attributes anyway:
> 
> > I prefer this syntax for the web3 object as it separates them and prevents collisions (while also letting you know what is global and what is based on cards)
> 
> ```
> web3.tokens.dataChanged = (oldTokens, updated, tokenIdCard) => {
>     document.getElementById(tokenIdCard).innerHTML = new Token(updated.token, updated.card).render();
> };
> 
> constructor(token, card) {
>         this.token = token;
>         this.card = card;
> }
> ```

@colourful-land am simply copying how it is written in the contract.

James-Sangalli commented:
@colourful-land please make another PR for other items to go into this branch. That way it can be easy to maintain. e.g. translation-activity -> aETH-activity. You can also rebase your PR branch with this one whenever you need without causing a problem in here.

hboon commented on examples/action/ETH/ETH.xml:
```diff
@@ -0,0 +1,243 @@
+
+        <ts:card type="activity" name="aETHMinted">
+            <ts:origins>
+                <ethereum:event type="Mint" filter="ownerAddress=${_fromMinter}"/>
```
Probably should be:

```
type="MintOnDeposit"
```

hboon commented:
@colourful-land @James-Sangalli BTW, there is one aspect that I don't think we covered in the TokenScript design for Activity:

In an activity card mapped from an event that was emitted at block `b`, what if we want to display the historical value at block `b` instead of `latest`?

A contrived example: "You donated 10 ETH, so you only have 2 ETH left". The donated amount is going to be part of the event, but the balance isn't.

Are there situations like that? Not wanting to design for this now, but just to highlight in case we encounter this for the PR.

hboon commented on examples/erc20/DAI/DAI.xml:
```diff
@@ -61,18 +94,192 @@ xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+
+        <ts:card type="activity" name="aETHRepaidFromAnotherUser">
+            <ts:origins>
+                <ethereum:event type="Repay" filter="&amp;(ownerAddress=${_user}&amp;!(ownerAddress=${_repayer})"/>
```
There's a syntax error in the filter, probably should be:

```
&amp;(ownerAddress=${_user}(!(ownerAddress=${_repayer})))
```

(I haven't tested it  yet)

hboon commented on examples/erc20/DAI/repay.en.js:
```diff
@@ -0,0 +1,46 @@
+
+    render() {
+        const decimals18 = 1e+18;
+        let daiDebt = this.props.borrowBalance / decimals18;
```
These few references to `this.props` should be `this.token` instead.

hboon commented on examples/erc20/DAI/repay.en.js:
```diff
@@ -0,0 +1,46 @@
+          <div id="inputBox">
+              <bold><h3>Repay DAI debt to AAVE</h3></bold>
+              <span><input id="repayAmount" type="number"></span>
+              <span><input id="debtHolder" type="text">${this.props.ownerAddress}</span>
```
`this.token` here too

hboon commented on examples/erc20/DAI/repay.en.js:
```diff
@@ -0,0 +1,46 @@
+    }
+}
+
+web3.tokens.dataChanged = (oldTokens, updatedTokens, tokenCardId) => {
```
```
web3.tokens.dataChanged = (oldTokens, updated, tokenIdCard) =&gt; {
    document.getElementById(tokenIdCard).innerHTML = new Token(updated.token, updated.card).render();
};
```

James-Sangalli replied:
CDATA wrapping handles this

hboon commented on examples/erc20/DAI/DAI.xml:
```diff
@@ -61,18 +94,192 @@ xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+
+        <ts:card type="activity" name="aETHRepaidFromAnotherUser">
+            <ts:origins>
+                <ethereum:event type="Repay" filter="&amp;(ownerAddress=${_user}(!(ownerAddress=${_repayer})))"/>
```
Needs the `<asnx:module>` type definition for the `Repay` event :)

colourful-land replied:
@James-Sangalli that and also the `contract="…"`

James-Sangalli replied:
dfe9de7

hboon commented on examples/erc20/aETH/aETH.xml:
```diff
@@ -0,0 +1,403 @@
+
+        <ts:card type="activity" name="aETHBorrowed">
+            <ts:origins>
+                <ethereum:event type="Borrow" filter="ownerAddress=${_user}"/>
```
There isn't a matching module type defined for this.

hboon replied:
`filter="ownerAddress=${_user}"`

The order the filter is written seems a bit strange to me. I thought this would be a bit more natural:

`filter="_user=${ownerAddress}"`

As in:

`filter="attribute=value"` rather than `filter="value=attribute"`

But I *think* it works syntactically (will have to test), so just an observation.

hboon replied:
I think the interest rate might be too good to be true :)

James-Sangalli replied:
@hboon changed, will check up on the interest rate later

hboon commented on examples/action/ETH/ETH.xml:
```diff
@@ -0,0 +1,243 @@
+    <ts:contract name="aETH" interface="erc20">
+        <ts:address network="1">0x3a3a65aab0dd2a17e3f1947ba16138cd37d08c04</ts:address>
+    </ts:contract>
+    <ts:contract name="ETH">
```
Can you make it specify that it's "erc20", like this? :

```
<ts:contract name="ETH" interface="erc20">
```

The only thing effects this has currently is to add the default actions for that `interface`. In this case the ERC20 send and receive actions, which behave the same way for native crypto in the app.

The "Deposit" action seems to show up and gets displayed

James-Sangalli replied:
@hboon it won't confuse that for an erc20 contract in terms of logic tho? I guess you will catch that in the code base

hboon replied:
Not at the moment. Setting it is only makes the send and receive button appear in that token screen shared by fungibles. When those 2 buttons are tapped, it will sort out of it’s an ERC20 or native crypto. 

Right after the train, we might to fix it though since this would signify inheritance of the base TokenScript.

hboon commented on examples/action/ETH/ETH.xml:
```diff
@@ -0,0 +1,243 @@
+            </ts:transaction>
+            <ts:view xml:lang="en">
+                <xhtml:style type="text/css">&style;</xhtml:style>
+                <xhtml:script type="text/javascript">&deposit.en;</xhtml:script>
```
The balance display is switched up in the view.

It's doing something like:

```
ETH Balance: ${aEthBalance}
aETH Balance: ${EthBalance}
```

hboon replied:
This attribute doesn't exist:

`${this.token.rate}`

hboon replied:
I pushed to the iOS branch again. so `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` should be accepted now.

colourful-land replied:
> I pushed to the iOS branch again. so `0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE` should be accepted now.

FYI That's

-  the identifier of ETH native currency in terms of **AAVE** (like when enquiring Lending Pool liquidity for ETH).

not sure if @James-Sangalli also intend to be

- the identifier of ETH native currency in terms of **TokenScript**

(I'm not for or against that decision, just FYI).

James-Sangalli replied:
@hboon I intend to use the null address in the card to be consistent with the realm storage. the Ee address is for aave purposes only.

hboon commented on examples/erc20/aETH/aETH.xml:
```diff
@@ -0,0 +1,403 @@
+
+        <ts:card type="activity" name="aETHRedeemed">
+            <ts:origins>
+                <ethereum:event type="Redeem" filter="ownerAddress=${_fromRedeemed}"/>
```
Probably:

```
<ethereum:event filter="ownerAddress=${_fromRedeemer}" type="Redeem"/>
```

James-Sangalli commented on examples/2020-06.mk:
```diff
@@ -1,5 +1,5 @@
 ifeq ($(TOKENSCRIPT_SCHEMA),)
-TOKENSCRIPT_SCHEMA=http://tokenscript.org/2020/06/tokenscript.xsd
+TOKENSCRIPT_SCHEMA=/Users/jamessangalli/Documents/projects/TokenScript/schema/tokenscript.xsd
```
Opps, will remove this later...

James-Sangalli replied:
@colourful-land can we put the new schema safely on the server?

hboon commented on examples/action/ETH/ETH.xml:
```diff
@@ -0,0 +1,243 @@
+
+        <ts:card type="activity" name="aETHMinted">
+            <ts:origins>
+                <ethereum:event type="MintOnDeposit" filter="ownerAddress=${_fromMinter}" contract="aETH"/>
```
Unfortunately, you'll have to switch the filters around so it becomes:

```
filter="_fromMinter=${ownerAddress}"
```

I'll check after the train leaves if we can also support the original way you wrote it.

hboon commented on examples/erc20/DAI/DAI.xml:
```diff
@@ -167,9 +188,9 @@ xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+        <ts:card type="activity" name="daiRepaidFromAnotherUser">
             <ts:origins>
-                <ethereum:event type="Repay" filter="&amp;(ownerAddress=${_user}(!(ownerAddress=${_repayer})))"/>
+                <ethereum:event type="Repay" filter="&amp;(ownerAddress=${_user}(!(ownerAddress=${_repayer})))" contract="aDAI"/>
```
This filter isn't supported unfortunately. Also might not be possible with fetching event logs (need to check after the train).

James-Sangalli replied:
Dropped for demo purposes

James-Sangalli commented on examples/action/ETH/ETH.xml:
```diff
@@ -100,6 +101,20 @@
         </ts:card>
 
         <ts:card type="activity" name="aETHMinted">
+            <ts:origins>
```
@colourful-land is the text that goes in here https://app.zeplin.io/project/5d0356be2f25b319dcc5944b/screen/5f101c19f9da398cca6382df defined a label in this block?

James-Sangalli replied:
No, this is the item view
